pkg/rotator/rotator.go

```diff
@@ -45,18 +45,42 @@ const (
 
 var crLog = logf.Log.WithName("cert-rotation")
 
-var vwhGVK = schema.GroupVersionKind{Group: "admissionregistration.k8s.io", Version: "v1beta1", Kind: "ValidatingWebhookConfiguration"}
+//WebhookType it the type of webhook, either validating/mutating webhook or a CRD conversion webhook
+type WebhookType int
 
-var crdGVK = schema.GroupVersionKind{Group: "apiextensions.k8s.io", Version: "v1beta1", Kind: "CustomResourceDefinition"}
+const (
+	//ValidatingWebhook indicates the webhook is a ValidatingWebhook
+	Validating WebhookType = iota
+	//MutingWebhook indicates the webhook is a MutatingWebhook
+	Mutating
+	//CRDConversionWebhook indicates the webhook is a conversion webhook
+	CRDConversion
+)
 
 var _ manager.Runnable = &CertRotator{}
 
 var restartOnSecretRefresh = false
 
+//WebhookInfo is used by the rotator to recieve info about resources to be updated with certificates
+type WebhookInfo struct {
+	//Name is the name of the webhook for a validating or mutating webhook, or the CRD name in case of a CRD conversion webhook
+	Name string
+	Type WebhookType
+}
+
 func init() {
 	flag.BoolVar(&restartOnSecretRefresh, "cert-restart-on-secret-refresh", false, "Kills the process when secrets are refreshed so that the pod can be restarted (secrets take up to 60s to be updated by running pods)")
 }
 
+func (w WebhookInfo) gvk() schema.GroupVersionKind {
+	t2g := map[WebhookType]schema.GroupVersionKind{
+		Validating:    schema.GroupVersionKind{Group: "admissionregistration.k8s.io", Version: "v1beta1", Kind: "ValidatingWebhookConfiguration"},
```

Comment:
should we also support v1?

Reply:
I think this will work fine for users using v1 as well, right? In theory, it's exactly the same object, just being "viewed" as either v1 or v1beta.

Reply:
Thanks @adrianludwin! This is very interesting. I just tested Gatekeeper by updating the validatingwebhookconfiguration resource to v1 and the `watch` and `get` are working as expected. Is [this](https://github.com/kubernetes-sigs/controller-runtime/blob/master/pkg/client/apiutil/apimachinery.go#L51) what you were referring to in terms of it will be "viewed" as either v1 or v1beta1?

Reply:
(Back at a real keyboard so I can give longer answers now)

AIUI, each K8s object has a "storage version" which is basically the union of all currently supported API versions. So one client can write a webhook config using v1, while the rotator can modify _the same object_ using v1beta1. The page I usually refer to is [this one](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api_changes.md#operational-overview). If you're adding a type via a CRD, that CRD can have multiple versions, _exactly one_ of which can be explicitly marked as the "storage version," and then you can write a webhook to convert between different versions.

I don't _think_ the link you copied is the most relevant one, since it's in `controller-runtime` (while I'm referring to a core K8s concept). I'm _guessing_ that line 73 is doing a bunch of extra work there, and picking the "best" version that's available for a given group/kind.

All of which is a long way of saying that yes: if GK uses v1 for its webhook configs and cert-controller uses v1beta1, everything should just work until a version of K8s comes along that no longer supports v1beta1, which [looks like it will be 1.22](https://github.com/kubernetes/kubernetes/issues/82021). So some time before 1.22 arrives, we'll switch cert-controller over.

We should probably wait until all the major hosted implementation have dropped 1.15 before we remove v1beta1 from cert-controller. As of today, 1.15 is still supported by [AKS](https://docs.microsoft.com/en-us/azure/aks/supported-kubernetes-versions), [EKS](https://docs.aws.amazon.com/eks/latest/userguide/kubernetes-versions.html), and [GKE](https://cloud.google.com/kubernetes-engine/docs/release-notes), so I'd recommend waiting a bit longer; hopefully they'll all drop 1.15 before 1.22 lands.

Does this all sound right?

Reply:
Ah, a more concise statement is [here](https://kubernetes.io/docs/concepts/overview/kubernetes-api/#api-groups-and-versioning):

> The API server may serve the same underlying data through multiple API version and handle the conversion between API versions transparently. All these different versions are actually representations of the same resource. For example, suppose there are two versions v1 and v1beta1 for the same resource. An object created by the v1beta1 version can then be read, updated, and deleted by either the v1beta1 or the v1 versions.

Reply:
Thanks for the explanation @adrianludwin! I thought it was something special the controller-manager is doing. But good to know that this done by the api server.  

> So some time before 1.22 arrives, we'll switch cert-controller over.

SGTM! I was not suggesting to switch to v1. I was just thinking to add v1 in case someone is using that to cover our basis.

Reply:
Great! You can tell that @yiqigao217 and I have been [thinking a lot about API versions](http://bit.ly/hnc_v1alpha2) lately...

So to sum up, I think we agree that this code is fine as-is 🙂

Reply:
> You can tell that @yiqigao217 and I have been thinking a lot about API versions lately...

Yes! Thanks for the explanation. TIL!

Yes this PR is great. Good work @mmirecki! I just left few nits for typos. but other than those, LGTM!